### ISSUE TITLE
yolo10 after converting to tflite not showing correct result when running thorugh tensorflow c++

### ISSUE THREAD

**Issue by Ashish2000L** [ultralytics/ultralytics#16451]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

I have trained and converted yolo10s.pt model to tflite so that I can run the same with tensorflow lite c++ apis. After predicting from yolo using 

```
!yolo predict task=detect model=/kaggle/working/runs/detect/train6/weights/best_saved_model/best_float16.tflite imgsz=640 half source="/kaggle/working/72c6e68215208ca6444071e881ca9dfe8c5ffafb.jpeg"
```

which is giving me best result as expected, but when using c++ tensorflow code which is giving way worse result then expected it is giving more like random results.

I have checked the other issues, stackoverflows and my code i.e input shapes, normalization and output shapes. Everything is looking good as of my knowledge.

I have also checked my nms and it is removing the invalid results correctly.

this is the image I am using to identify the number in it:-
![img](https://github.com/user-attachments/assets/e5c3a996-b1fc-45a7-8ab1-321823e360c2)

here is the model I am using, remove the .txt extention from the end
[model.tflite.txt](https://github.com/user-attachments/files/17109498/model.tflite.txt)

this is the result I am getting after predicting from tensorflow c++ apis
![bound_img](https://github.com/user-attachments/assets/00596730-edbb-4309-8f07-da3ff7579dd8)

here is the image when using above yolo code for predicting the image which is giving prefect as required:-
<img width="413" alt="image" src="https://github.com/user-attachments/assets/30b67678-e6db-4bde-938b-22eb3ef4ebd7">

### Environment

``` 
ultralytics-8.2.99
python 3.12
```

### Minimal Reproducible Example

```
#include <iostream>
#include <tuple>
#include <tensorflow/lite/interpreter.h>
#include <tensorflow/lite/model.h>
#include <tensorflow/lite/kernels/register.h>
#include <opencv2/opencv.hpp>


float IoU(const std::tuple<float, float, float, float, float>& box1, const std::tuple<float, float, float, float, float>& box2) {  
    float x1 = std::get<0>(box1), y1 = std::get<1>(box1), w1 = std::get<2>(box1), h1 = std::get<3>(box1);
    float x2 = std::get<0>(box2), y2 = std::get<1>(box2), w2 = std::get<2>(box2), h2 = std::get<3>(box2);

    float x_left = std::max(x1, x2);
    float y_top = std::max(y1, y2);
    float x_right = std::min(x1 + w1, x2 + w2);
    float y_bottom = std::min(y1 + h1, y2 + h2);

    if (x_right < x_left || y_bottom < y_top) {
        return 0.0;
    }

    float intersection_area = (x_right - x_left) * (y_bottom - y_top);
    float box1_area = w1 * h1;
    float box2_area = w2 * h2;
    
    float union_area = box1_area + box2_area - intersection_area;
    return intersection_area / union_area;
}

std::vector<std::tuple<float, float, float, float, float>> applyNMS(
    std::vector<std::tuple<float, float, float, float, float>>& detections, float iou_threshold) {

    std::sort(detections.begin(), detections.end(), [](const auto& d1, const auto& d2) {
        return std::get<4>(d1) > std::get<4>(d2);  // Sort by confidence
    });

    std::vector<std::tuple<float, float, float, float, float>> nms_result;
    std::vector<bool> suppressed(detections.size(), false);

    for (size_t i = 0; i < detections.size(); ++i) {
        if (suppressed[i]) continue;
        nms_result.push_back(detections[i]);

        for (size_t j = i + 1; j < detections.size(); ++j) {
            if (IoU(detections[i], detections[j]) > iou_threshold) {
                suppressed[j] = true;
            }
        }
    }
    return nms_result;
}


int main() {

	tflite::StderrReporter error_reporter;
	std::unique_ptr<tflite::Interpreter> interpreter;

	std::unique_ptr<tflite::impl::FlatBufferModel> model = tflite::impl::FlatBufferModel::BuildFromFile("model.tflite", &error_reporter);

	tflite::ops::builtin::BuiltinOpResolver resolver;
	tflite::InterpreterBuilder builder(*model, resolver);
	builder(&interpreter);

	if (!interpreter) {
		std::cerr << "Failed to create interpreter" << std::endl;
		return -1;
	}

	TfLiteStatus status = interpreter->AllocateTensors();

	if (status != kTfLiteOk) {
		std::cerr << "Failed to allocate tensors."<<status << std::endl;
		return -1;
	}
	
	const TfLiteTensor* input_tensor = interpreter->input_tensor(0);  // Assume 0th input for simplicity
	const TfLiteIntArray* dims = input_tensor->dims;

	int input_batch = dims->data[0];    // Batch size (1)
	int input_height = dims->data[1];   // Input image height
	int input_width = dims->data[2];    // Input image width
	int input_channels = dims->data[3]; // Number of channels (e.g., 3 for RGB)
	
        std::cout<<"accepted channel: "<<input_channels<<std::endl;

	cv::Mat img = cv::imread("img.jpeg");

	if (img.empty()) {
        	std::cerr << "Error: Could not load image!" << std::endl;
        	return -1;
	}	

	cv::Mat resized_img;
	cv::resize(img, resized_img, cv::Size(input_width, input_height));
	cv::imwrite("resize_img1.png", resized_img);

	//normalize
	resized_img.convertTo(resized_img, CV_32FC3, 1.0 / 255.0);

	std::cout << "Resized Image Shape: " << resized_img.rows << " x " << resized_img.cols 
          << " x " << resized_img.channels() << std::endl;
	
	// Copy the image to the input tensor
	float* input = interpreter->typed_input_tensor<float>(0);
	std::memcpy(input, resized_img.data, resized_img.total() * resized_img.elemSize());
        
	status = interpreter->Invoke();
	
	if (status != kTfLiteOk) {
		std::cerr << "Failed to invoke tensors." << status << std::endl;
		return -1;
	}
	float* output = interpreter->typed_output_tensor<float>(0);

	auto output_tensor = interpreter->tensor(interpreter->outputs()[0]);

	const auto tensor_shape = output_tensor->dims;
	
	int steps = *((output_tensor->dims)[2].data);

	int num_detections = *((output_tensor->dims)[1].data); 

	std::vector<std::tuple<float, float, float, float, float>> detections;

	for (int i = 0; i < num_detections; i+=steps) {
	    float x = output[i + 0];
	    float y = output[i+ 1];
	    float width = output[i+ 2];
	    float height = output[i + 3];
	    float confidence = output[i + 4];

	    // Print the coordinates if confidence is above a certain threshold
	    if (confidence > 0.5) { // Adjust threshold as needed
				float cx = x;
                float cy = y;
                float w = width;
                float h = height;
                float x1 = cx - (w / 2.0);
                float y1 = cy - (h / 2.0);
                float x2 = cx + (w / 2.0);
                float y2 = cy + (h / 2.0);

                if (x1 < 0 || x1 > 1.0) continue;
                if (y1 < 0 || y1 > 1.0) continue;
                if (x2 < 0 || x2 > 1.0) continue;
                if (y2 < 0 || y2 > 1.0) continue;

		detections.push_back(std::make_tuple(cx, cy, width, height, confidence));
/*        	std::cout << "Detected object " << i + 1 << ": "
	                  << "Coordinates (" << x_min << ", " << y_min << ") - ("
	                  << x_max << ", " << y_max << ") with confidence " << confidence << std::endl;
			  */
	    }else {
		//std::cout<<"confidance is too low"<<confidence<<std::endl;
	    }
	}
	
	std::cout<<"analyzing image"<<std::endl;

	float iou_threshold = 0.5;  // Tune this threshold as needed
	auto best_detections = applyNMS(detections, iou_threshold);


	if (!best_detections.empty()) {
		for(int i=0; i<best_detections.size(); i++){
	    auto best_detection = best_detections[i];
	    float x = std::get<0>(best_detection);
	    float y = std::get<1>(best_detection);
	    float width = std::get<2>(best_detection);
	    float height = std::get<3>(best_detection);
	    float confidence = std::get<4>(best_detection);

	    std::cout<<"found number with confidence: "<<confidence<<std::endl;

	    int original_width = img.cols;
	    int original_height = img.rows;

	    int resized_width=resized_img.cols;
	    int resized_height = resized_img.rows;

		float scale_w = 1;
		float scale_h = 1;
	    
	    x *=scale_w;
	    y *= scale_h;
	    width *= scale_w;
	    height *= scale_h;

	    
	    float topLeftX = x*original_width;//(x-(width/2))*resized_width;
	    float topLeftY = y*original_height;//(y-(height/2))*resized_height;

	    cv::Rect boundingBox(topLeftX, topLeftY, width*original_width, height*original_height);
	

	    cv::Scalar color(255, 0, 0);
	    int thickness = 2;
	    cv::rectangle(img, boundingBox, color, thickness);
	}

	    cv::imwrite("bound_img.png", img);
	
	    std::cout << "Image saved" << std::endl;
	}else{
		std::cout<<"unable to find the number"<<std::endl;
	}
	return 0;
}
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Ashish2000L, thank you for reaching out and for your interest in Ultralytics 🚀! This is an automated response. An Ultralytics engineer will assist you soon.

For your issue, we recommend checking out our [Docs](https://docs.ultralytics.com) for guidance, including many usage examples. 

If this is a 🐛 Bug Report, please ensure you've provided a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us debug it. It looks like you've already included an MRE, which is great!

If you have a custom training ❓ Question, share more details like dataset examples and logs, and double-check you're adhering to our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

### Upgrade

Ensure you're using the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

### Environments

YOLOv8 works best in any of these verified environments:

- **Notebooks** with GPUs: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud**: See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon**: See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker**: See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

### Community

Join our vibrant community on [Discord](https://ultralytics.com/discord) 🎧 for real-time discussions, or participate in threads on our [Subreddit](https://reddit.com/r/ultralytics)!

### CI Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

This badge indicates the status of our continuous integration (CI) tests, which ensure the reliability of all YOLOv8 functionalities. If it's green, all tests are passing.

Stay tuned for further updates, and thank you for using Ultralytics! 😊

**Comment by Y-T-G** [comment]
Seems like an issue with scaling.

**Comment by Ashish2000L** [comment]
> Seems like an issue with scaling.
> 
> YOLOv10 outputs xyxy format and not xywh format.

can you share doc where I can found these info like output and other stuff, I was not able to get detailed doc for yolo10

**Comment by Y-T-G** [comment]
Not a bug, since it's an issue with your custom code and not ultralytics.

**Comment by Ashish2000L** [comment]
> Not a bug, since it's an issue with your custom code and not ultralytics.

if you can share the doc that will be really helpful for yolo10 where I can find the output structure of yolo10

**Comment by Y-T-G** [comment]
You need to change the postprocessing function.

The model outputs x1, y1, x2, y2. (top left and bottom right) coordinates.

You scale it by `x1 = x1 * original_width/resized_width`, `y1 = y1 * original_height/resized_height` and the same for x2 and y2.

**Comment by Ashish2000L** [comment]
> You need to change the postprocessing function.
> 
> The model outputs x1, y1, x2, y2. (top left and bottom right) coordinates.
> 
> You scale it by `x1 = x1 * original_width/resized_width`, `y1 = y1 * original_height/resized_height` and the same for x2 and y2.

as for the speed, yolo10s is also taking around 3-4sec after converting to tflite. is it a export issue or speed is expected to reduce when exporting or using tensorflow lite?

**Comment by Y-T-G** [comment]
Depends on your CPU. TFLite runs on the CPU, not GPU.
